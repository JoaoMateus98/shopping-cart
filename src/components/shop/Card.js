import AddToCart from "./AddToCart";

const Card = ({ game }) => {
  const { name, id, image, rating, price } = createGameObject(game);
  const gameObject = createGameObject(game);

  return (
    <section data-testid={id} className="card-container">
      <div className="card">
        <img
          src={image}
          alt={"Image of ".concat(name)}
          width="256"
          height="144"
        />
        <h2>{name}</h2>
        <p>{"Metacritic Score: ".concat(rating)}</p>
        <div className="add-to-cart-container">
          <p>${price}</p>
          <AddToCart currentGame={gameObject} />
        </div>
      </div>
    </section>
  );
};

export function createGameObject(game) {
  const name = game.name;
  const id = game.id;
  const image = game.background_image;
  const rating = game.metacritic;
  const price = (game.rating * 10).toFixed(2); // fetched item doesnt have a real price

  return {
    name: name,
    id: id,
    image: image,
    rating: rating,
    price: price,
  };
}

export default Card;
