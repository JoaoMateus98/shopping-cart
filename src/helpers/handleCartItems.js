export function isAlreadyInCart(cartItems, currentGame) {
  let alreadyInCart = false;

  cartItems.forEach((item) => {
    if (item.id === currentGame.id) {
      alreadyInCart = true;
    }
  });

  return alreadyInCart;
}

export function incrementCart(cartItems, currentGame, setCartItems) {
  let cartCopy = [...cartItems];
  cartCopy.forEach((item) => {
    if (item.id === currentGame.id) {
      item.quantity++;
      return;
    }
  });

  setCartItems([...cartCopy]);
}

export function decrementCart(cartItems, currentGame, setCartItems) {
  const cartCopy = [...cartItems];

  cartCopy.forEach((item) => {
    if (item.id === currentGame.id) {
      if (item.quantity === 1) {
        const removedItemArray = cartCopy.filter(
          (curItem) => item.id !== curItem.id
        );
        setCartItems([...removedItemArray]);
      } else {
        item.quantity--;
      }
    }
  });

  setCartItems([...cartCopy]);
}

export function countTotalQuantity(cartItems) {
  const total = cartItems.reduce(
    (accumulator, item) => accumulator + item.quantity,
    0
  );
  return total;
}

export function getItemTotal(cartItem) {
  return cartItem.price * cartItem.quantity;
}

export function getGrandTotalPrice(cartItems) {
  const total = cartItems.reduce(
    (accumulator, curItem) => accumulator + curItem.price * curItem.quantity,
    0
  );

  return total;
}
