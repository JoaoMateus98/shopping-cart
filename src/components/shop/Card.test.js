import { render, screen } from "@testing-library/react";
import { testGameObject } from "../../helpers/testGameObject";
import "@testing-library/jest-dom";
import Card, { createGameObject } from "./Card";

test("create object with all need properties", () => {
  expect(createGameObject(testGameObject)).toEqual({
    name: testGameObject.name,
    id: testGameObject.id,
    image: testGameObject.background_image,
    rating: testGameObject.metacritic,
    price: (testGameObject.rating * 10).toFixed(2),
  });
});

describe("creates card element", () => {
  test("renders correctly", () => {
    const { container } = render(<Card game={testGameObject} />);

    expect(container).toMatchSnapshot();
  });

  test("display game name", () => {
    render(<Card game={testGameObject} />);

    expect(
      screen.getByRole("heading", { name: testGameObject.name })
    ).toHaveTextContent(testGameObject.name);
  });

  test("card element contains all child elements", () => {
    render(<Card game={testGameObject} />);

    const cardContainer = screen.getByTestId(testGameObject.id);
    const cardImage = screen.getByRole("img");
    const cardName = screen.getByRole("heading", {
      name: testGameObject.name,
    });
    const rating = screen.getByText(
      "Metacritic Score: ".concat(testGameObject.metacritic)
    );
    const price = screen.getByText(
      "$".concat((testGameObject.rating * 10).toFixed(2))
    );
    const addToCartButton = screen.getByRole("button");

    expect(cardContainer).toContainElement(cardImage);
    expect(cardContainer).toContainElement(cardName);
    expect(cardContainer).toContainElement(rating);
    expect(cardContainer).toContainElement(price);
    expect(cardContainer).toContainElement(addToCartButton);
  });
});
