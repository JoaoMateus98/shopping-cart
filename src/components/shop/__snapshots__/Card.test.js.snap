// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates card element renders correctly 1`] = `
<div>
  <section
    class="card"
    data-testid="3498"
  >
    <img
      alt="Image of Grand Theft Auto V"
      height="144"
      src="https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg"
      width="256"
    />
    <h2>
      Grand Theft Auto V
    </h2>
    <p>
      Metacritic Score: 91
    </p>
    <p>
      44.70
    </p>
    <button>
      Add To Cart
    </button>
     // TODO: replace with AddToCart component
  </section>
</div>
`;
